v3/src/components/graph/adornments/adornment.tsx

```diff
@@ -12,35 +17,63 @@ interface IProps {
   yCategories: string[] | number[]
 }
 
-export const Adornment = ({ adornment, index, xCategories, yCategories}: IProps) => {
+export const Adornment = observer(function Adornment({ adornment, index, xCategories, yCategories}: IProps) {
   const graphModel = useGraphModelContext(),
     layout = useGraphLayoutContext(),
     subPlotWidth = xCategories.length > 0
                      ? layout.plotWidth / xCategories.length
                      : layout.plotWidth,
     subPlotHeight = yCategories.length > 0
                       ? layout.plotHeight / yCategories.length
-                      : layout.plotHeight
+                      : layout.plotHeight,
+    isFadeInComplete = useRef(false),
+    isFadeOutComplete = useRef(false)
+
+  useEffect(function fadeInCleanup() {
+    isFadeInComplete.current = adornment.isVisible
+    isFadeOutComplete.current = !adornment.isVisible
+  }, [adornment.isVisible])
 
   // The instanceKey is used by the adornment model to uniquely identify an instance of the adornment in its
   // map of instances. If there is only one instance of the adornment, the instanceKey is an empty string.
-  const instanceKey = adornment.instanceKey(xCategories, yCategories, index)
-  // The adornmentKey is a unique value used for React's key prop. We can't use the instanceKey because that
-  // value may be duplicated if there are multiple types of adornments active on the graph.
-  const adornmentKey = `${adornment.id}${instanceKey ? `-${instanceKey}` : ''}`
+  const instanceKey = adornment.instanceKey(xCategories, yCategories, index) ?? ''
+  const classFromKey = adornment.classNameFromKey(instanceKey)
+  // The adornmentKey is a unique value used for React's key prop and for the adornment wrapper's HTML ID.
+  // We can't use the instanceKey because that value may be duplicated if there are multiple types of
+  // adornments active on the graph.
+  const adornmentKey = `${adornment.id}${instanceKey ? `-${classFromKey}` : ''}`
   const componentInfo = getAdornmentComponentInfo(adornment.type)
   if (!componentInfo) return null
   const { Component } = componentInfo
+
+  const adornmentWrapperClass = clsx(
+    'adornment-wrapper',
+    `${adornmentKey}-wrapper`,
+    {
+      'fadeIn': adornment.isVisible && !isFadeInComplete.current,
+      'fadeOut': !adornment.isVisible && !isFadeOutComplete.current,
+      'hidden': !adornment.isVisible && isFadeOutComplete.current
+    }
+  )
```

Comment:
@emcelroy, this code starts to fail after upgrading mobx-react-lite from v3 to v4.

I think that rendering based on the current ref value should be generally considered a bug, as updates to refs won't trigger re-renders and it's kinda random what we'll see in the DOM in the end.

Also, what's the purpose of `isFadeInComplete` and `isFadeOutComplete`? I can see they're updated once in the useEffect, but never again.